File: app/src/main/java/com/hundaol/ethiopic/views/LabelStamp.java
package com.hundaol.ethiopic.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.text.TextPaint;

import com.hundaol.ethiocal.R;

import timber.log.Timber;

/**
 * Created by john.pirie on 2017-04-30.
 */

public class LabelStamp {

    private final Context context;
    private final CalendarViewModel viewModel;
    private final RectF bounds = new RectF();

    public final Paint backPaint;
    public final Paint forePaint;
    public final TextPaint textPaint;

    public int jdn;

    public LabelStamp(Context context, CalendarViewModel viewModel) {
        this.context = context;
        this.viewModel = viewModel;

        backPaint = new Paint();
        backPaint.setStyle(Paint.Style.FILL);
        backPaint.setColor(context.getResources().getColor(R.color.label_background));

        forePaint = new Paint();
        forePaint.setStyle(Paint.Style.STROKE);
        forePaint.setColor(context.getResources().getColor(R.color.label_foreground));
        forePaint.setStrokeWidth(context.getResources().getDimensionPixelSize(R.dimen.label_border_width));

        textPaint = new TextPaint();
        textPaint.setTextSize(context.getResources().getDimensionPixelSize(R.dimen.label_text_size));
        textPaint.setStyle(Paint.Style.STROKE);
        textPaint.setColor(context.getResources().getColor(R.color.label_text));
    }

    public void setJdn(int jdn) {
        this.jdn = jdn;
        bounds.set(viewModel.boundsForLabel(jdn));
    }

    public void stamp(Canvas canvas) {
        canvas.drawRect(bounds, backPaint);

        canvas.drawRect(bounds, forePaint);

        String s =  new String(viewModel.getCal().getMonthName(jdn)) + ", " + viewModel.getCal().getYear(jdn);
        float tw = textPaint.measureText(s);
        float th = textPaint.getTextSize();
        canvas.rotate(90, bounds.centerX(), bounds.centerY());
        canvas.drawText(s, bounds.left + (bounds.width() - tw) / 2.0f, bounds.top + (bounds.height() + th) / 2.0f, textPaint);
        canvas.rotate(-90, bounds.centerX(), bounds.centerY());
    }
}

File: app/src/main/java/com/hundaol/ethiopic/stamps/LabelStamp.kt
package com.hundaol.ethiopic.stamps

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.text.TextPaint
import com.hundaol.ethiopic.cal.GregorianCal
import com.hundaol.ethiopic.cal.ICal
import com.hundaol.ethiopic.domain.DateModel
import com.hundaol.ethiopic.domain.ColorModel
import com.hundaol.ethiopic.views.calendar.CalendarViewModel

/**
 * Created by john.pirie on 2017-06-06.
 */

class LabelStamp(var jdn: Int = 0,
                 var cal: ICal = GregorianCal.INSTANCE) {

    var viewModel = CalendarViewModel()
    var dateModel: DateModel = DateModel.default

    var colorModel: ColorModel = ColorModel.default

    val bounds: RectF
        get() = viewModel.boundsForLabel(dateModel, cal, jdn)

    val backPaint = Paint()
    val textPaint = TextPaint()
    val gridPaint = Paint()

    init {
        backPaint.style = Paint.Style.FILL_AND_STROKE

        textPaint.style = Paint.Style.STROKE

        gridPaint.style = Paint.Style.STROKE
        gridPaint.strokeWidth = 4.0f
    }

    fun stamp(canvas: Canvas) {
        val bounds = this.bounds

        backPaint.color = colorModel.backgroundColorForMonth(cal, jdn)
        canvas.drawRect(viewModel.boundsForLabel(dateModel, cal, jdn), backPaint)

        textPaint.color = colorModel.foregroundColorForLabel(cal, jdn)
        textPaint.textSize = viewModel.getTextSizeForLabel(dateModel, cal, jdn)

        val s = String(cal.getMonthName(jdn)) + ", " + cal.getYear(jdn)
        val tw = textPaint.measureText(s)
        val th = textPaint.getTextSize()
        canvas.rotate(90.0f, bounds.centerX(), bounds.centerY())
        canvas.drawText(s, bounds.left + (bounds.width() - tw) / 2.0f, bounds.top + (bounds.height() + th) / 2.0f, textPaint)
        canvas.rotate(-90.0f, bounds.centerX(), bounds.centerY())

        gridPaint.color = colorModel.gridColor(cal, jdn)
        canvas.drawRect(bounds, gridPaint)
    }
}
